x/cvm/keeper/state.go

```diff
@@ -97,10 +97,25 @@ func (s *State) UpdateAccount(updatedAccount *acm.Account) error {
 	} else {
 		cvmCode = types.NewCVMCode(types.CVMCodeTypeEVMCode, updatedAccount.EVMCode)
 	}
-	s.store.Set(types.CodeStoreKey(updatedAccount.Address), s.cdc.MustMarshalBinaryBare(&cvmCode))
-	err := s.bk.SetBalances(s.ctx, address, sdk.Coins{sdk.NewInt64Coin(s.sk.BondDenom(s.ctx), int64(updatedAccount.Balance))})
-	if err != nil {
-		return err
+	s.store.Set(types.CodeStoreKey(updatedAccount.Address), s.cdc.MustMarshal(&cvmCode))
+	oldBalance := s.bk.GetBalance(s.ctx, address, s.sk.BondDenom(s.ctx))
+	newBalance := sdk.NewInt64Coin(s.sk.BondDenom(s.ctx), int64(updatedAccount.Balance))
+	if newBalance.Amount.GT(oldBalance.Amount) {
+		coins := sdk.Coins{newBalance.Sub(oldBalance)}
+		if err := s.bk.MintCoins(s.ctx, types.ModuleName, coins); err != nil {
+			return err
+		}
+		if err := s.bk.SendCoinsFromModuleToAccount(s.ctx, types.ModuleName, address, coins); err != nil {
+			return err
+		}
+	} else if newBalance.Amount.LT(oldBalance.Amount) {
+		coins := sdk.Coins{oldBalance.Sub(newBalance)}
+		if err := s.bk.SendCoinsFromAccountToModule(s.ctx, address, types.ModuleName, coins); err != nil {
+			return err
+		}
+		if err := s.bk.BurnCoins(s.ctx, types.ModuleName, coins); err != nil {
+			return err
+		}
```

Comment:
all cvm executions will always have input coins equal to the output coins. I'm not sure if this is the best way to handle this. Let's see what other alternatives we have instead of minting and burning on every account balance updates.